scene/gui/control.cpp

```diff
@@ -1907,19 +1907,43 @@ Control::FocusMode Control::get_focus_mode() const {
 	return data.focus_mode;
 }
 
+void Control::set_focus_layer(int p_layer_id) {
+	// If this Control has focus, then lets find the next valid Control to focus before switching layers.
+	if (has_focus()) {
+		Control *next = find_next_valid_focus();
+
+		if (next && next != this) {
+			next->grab_focus(false);
+		} else {
+			release_focus();
+		}
+	}
+
+	data.focus_layer = p_layer_id;
+}
+
+int Control::get_focus_layer() const {
+	return data.focus_layer;
+}
+
 bool Control::has_focus() const {
 	return is_inside_tree() && get_viewport()->_gui_control_has_focus(this);
 }
 
-void Control::grab_focus() {
-	ERR_FAIL_COND(!is_inside_tree());
+bool Control::grab_focus(bool p_auto_set_viewport_active_focus_layer) {
+	ERR_FAIL_COND_V(!is_inside_tree(), false);
 
 	if (data.focus_mode == FOCUS_NONE) {
 		WARN_PRINT("This control can't grab focus. Use set_focus_mode() to allow a control to get focus.");
```

Comment:
I wonder if this warning is still necessary if the method returns bool 🤔
(up to discussion, don't change it yet)

Reply:
Yeah, not sure. TBH I'm not sure how a `is_inside_tree()` would fail. Maybe if you dynamically created a Control and didn't parent it to anything?